pkg/tecdsa/dkg/dkg.go

```diff
@@ -85,6 +86,101 @@ func (e *Executor) Execute(
 	return finalizationState.result(), endBlockNumber, nil
 }
 
+// StartedDKGEvent represents a DKG start event.
+type StartedDKGEvent struct {
+	Seed        *big.Int
+	BlockNumber uint64
+}
+
+// ResultSubmissionEvent represents a DKG result submission event. It is emitted
+// after a submitted DKG result is positively validated on the chain. It contains
+// the index of the member who submitted the result and a final public key of
+// the group.
+type ResultSubmissionEvent struct {
+	MemberIndex         uint32
+	GroupPublicKeyBytes []byte
+	Misbehaved          []uint8
+
+	BlockNumber uint64
+}
+
+type ResultSubmitter interface {
+	// SubmitDKGResult sends DKG result to a chain, along with signatures over
+	// result hash from group participants supporting the result.
+	// Signatures over DKG result hash are collected in a map keyed by signer's
+	// member index.
+	SubmitDKGResult(
+		participantIndex group.MemberIndex,
+		result *Result,
+		signatures map[group.MemberIndex][]byte,
+	) error
+	// IsGroupRegistered checks if group with the given public key is registered
+	// on-chain.
+	IsGroupRegistered(groupPublicKey []byte) (bool, error)
+	// OnDKGResultSubmitted registers a callback that is invoked when an on-chain
+	// notification of a new, valid submitted result is seen.
+	OnDKGResultSubmitted(
+		func(event *ResultSubmissionEvent),
+	) subscription.EventSubscription
+}
+
+type ResultSigner interface {
+	// Signing returns the chain's signer.
+	Signing() chain.Signing
+	// CalculateDKGResultHash calculates 256-bit hash of DKG result in standard
+	// specific for the chain. Operation is performed off-chain.
+	CalculateDKGResultHash(result *Result) (ResultHash, error)
+}
+
+type ResultHandler interface {
+	ResultSigner
+	ResultSubmitter
+}
+
+// Publish signs the DKG result for the given group member, collects signatures
+// from other members and verifies them, and submits the DKG result.
+func Publish(
+	logger log.StandardLogger,
+	memberIndex group.MemberIndex,
+	membershipValidator *group.MembershipValidator,
+	submissionConfig *SubmissionConfig,
+	result *Result,
+	channel net.BroadcastChannel,
+	resultHandler ResultHandler,
+	blockCounter chain.BlockCounter,
+	startBlockHeight uint64,
+) error {
```

Comment:
Let's follow the same order of parameters as in `Execute`.

Reply:
Done in dea8a0788aeddd585ecd96529e2edfa770d4b87c.